File: array.py
class ArrayList:
    def __init__(self):
        self.data = []      # Python list used as dynamic array
        self.size = 0

    # Append at the end
    def append(self, value):
        self.data.append(value)
        self.size += 1

    # Insert at position
    def insert_at(self, position, value):
        if position < 0 or position > self.size:
            print("Position out of range")
            return
        self.data.insert(position, value)
        self.size += 1

    # Delete at position
    def delete_at(self, position):
        if position < 0 or position >= self.size:
            print("Position out of range")
            return
        self.data.pop(position)
        self.size -= 1

    # Print list
    def print_list(self):
        print("ArrayList:", self.data)

    # Get length
    def get_size(self):
        return self.size

# =====================
if __name__ == "__main__":
    arr = ArrayList()

    arr.append(10)
    arr.append(20)
    arr.append(30)
    arr.insert_at(1, 15)

    arr.print_list()   # [10, 15, 20, 30]

    arr.delete_at(2)
    arr.print_list()   # [10, 15, 30]

    print("Size:", arr.get_size())

File: array.java
import java.util.Arrays;

public class DynamicArray {

    private int[] data;
    private int size;

    // Constructor
    public DynamicArray(int capacity) {
        data = new int[capacity];
        size = 0;
    }

    // Default constructor
    public DynamicArray() {
        this(2);
    }

    // Resize when full
    private void resize() {
        data = Arrays.copyOf(data, data.length * 2);
    }

    // Add at end
    public void append(int value) {
        if (size == data.length) {
            resize();
        }
        data[size++] = value;
    }

    // Insert at position
    public void insertAt(int pos, int value) {
        if (pos < 0 || pos > size) {
            System.out.println("Invalid position");
            return;
        }

        if (size == data.length) {
            resize();
        }

        for (int i = size; i > pos; i--) {
            data[i] = data[i - 1];
        }

        data[pos] = value;
        size++;
    }

    // Delete at position
    public void deleteAt(int pos) {
        if (pos < 0 || pos >= size) {
            System.out.println("Invalid position");
            return;
        }

        for (int i = pos; i < size - 1; i++) {
            data[i] = data[i + 1];
        }

        size--;
    }

    // Get element
    public int get(int pos) {
        if (pos < 0 || pos >= size) {
            System.out.println("Invalid index");
            return -1;
        }
        return data[pos];
    }

    // Update element
    public void update(int pos, int value) {
        if (pos < 0 || pos >= size) {
            System.out.println("Invalid index");
            return;
        }
        data[pos] = value;
    }

    // Search for value
    public int search(int value) {
        for (int i = 0; i < size; i++) {
            if (data[i] == value) {
                return i;
            }
        }
        return -1;
    }

    // Print array
    public void print() {
        System.out.print("Array: ");
        for (int i = 0; i < size; i++) {
            System.out.print(data[i] + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        DynamicArray arr = new DynamicArray(2);

        arr.append(10);
        arr.append(20);
        arr.append(30); // resize happens

        arr.print();

        arr.insertAt(1, 99);
        arr.print();

        System.out.println("Element at 2: " + arr.get(2));

        arr.update(1, 55);
        arr.print();

        arr.deleteAt(0);
        arr.print();

        int pos = arr.search(30);
        System.out.println("Search 30 found at: " + pos);
    }
}
